lib/dialects/abstract/registry.js

```diff
@@ -0,0 +1,36 @@
+
+const _registries = {};
+
+/**
+ * Register dialect implementation.
+ *
+ * @param {AbstractDialect} [DialectClass] dialect implemetation as subclass of AbstractDialect.
+ */
+function registerDialect( DialectClass ) {
+  _registries[DialectClass.getDialectName()] = DialectClass;
```

Comment:
We can throw if `(_registries[dialectName] != null && _registries[dialectName] !== DialectClass)` to ensure a user isn't accidentally registering the same dialect with two different implementations.

Reply:
> > Do I understand it correct that the main entry file for my dialect would call
> > ```
> > import {registerDialect} from 'sequelize'
> > 
> > registerDialect(MyDialectClass)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ?
> 
> My 1st approach is the dialect owner call it as u said.
> 
> But @ephys suggest we leave that responsibility to the user. So for user they might need to do it in the place they use that dialect:
> 
> ```js
> import Sequelize from 'sequelize';
> import myDialect from 'my-dialect';
> 
> //...
> Sequelize.registerDialect(myDialect); 
> //...
> ```
> 
> I am fine for either of them.

I think I would like to have something like this:

```
import Postgres from '@sequelize/postgres';
import Sequelize from '@sequelize/core';

const sequelize = new Sequelize(/* whatever */, { dialect: Postgres })
```

And now that I write it, I wonder what we need the registry for?!

Reply:
Sequelize supports database URIs (`new Sequelize('mysql://localhost:3306/database')`), it would need to use the registry to select the corresponding Dialect class.

It also keeps backward compatibility with specifying the dialect as string (given this is targeting v6) (`new Sequelize(/* whatever */, { dialect: 'pg' })`)

But we should also support passing the Dialect class as a parameter, bypassing the registry. (`new Sequelize(/* whatever */, { dialect: Postgres })`)

Reply:
Oh is this supposed to work in v6? I thought it's one of the early things for v7?

Reply:
I assumed it was v6 because of the backward compatibility but i'm not sure whether this was opened before the branch rename.

If this is targetting v7:

I'd vote to not break accepting URIs even but we could change the API to something that looks like this:

```typescript
new Sequelize(/* options */, { dialect: PostgresDialect });

new Sequelize('mysql://localhost:3306/database', {
   // array of dialects, one of them must support `mysql:`
  dialect: [MysqlDialect, PostgresDialect],
});
```

Reply:
But what is your use case to add more than one dialect? Do we currently support something like this where you connect with one sequelize instance to different databases?

Reply:
-->

```
new Sequelize(/* options */, { dialect: PostgresDialect });
new Sequelize('mysql://localhost:3306/database', {
   // array of dialects, one of them must support `mysql:`
  dialect: MysqlDialect,
});
```

Reply:
I honestly do not have one. I went with an array to mimic the current behavior where the protocol defines the dialect. The only use-case I could see would be to have the ability to let the config define the protocol. It's not a good use case and it's achievable if the user wants it anyway.

In sumary, this sounds good to me:

```typescript
new Sequelize(database, username, password, { dialect: PostgresDialect });
new Sequelize('mysql://localhost:3306/database', { dialect: MysqlDialect });
```

Reply:
💯 :)

Reply:
> new Sequelize(/* options */, { dialect: PostgresDialect });

That does not help much with the use case of specifying the dialect as an environment variable for tests. Unless ofc, the proper dialect is loaded in `test/supports.js`

Reply:
we can pass the right dialect based on the environment variable :)